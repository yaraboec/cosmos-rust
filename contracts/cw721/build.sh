#!/bin/bash
set -e && RUSTFLAGS='-C link-arg=-s' cargo wasm