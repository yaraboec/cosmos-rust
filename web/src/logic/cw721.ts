import { SigningCosmWasmClient } from "@cosmjs/cosmwasm-stargate";
import { Token } from "./types";

export class Cw721Contract {
  contractAddress: string;
  signingClient: SigningCosmWasmClient;

  constructor(contractAddress: string, signingClient: SigningCosmWasmClient) {
    this.contractAddress = contractAddress;
    this.signingClient = signingClient;
  }

  static async instantiate(
    senderAddress: string,
    signingClient: SigningCosmWasmClient,
    codeId: number
  ) {
    const { contractAddress } = await signingClient.instantiate(
      senderAddress,
      codeId,
      {
        name: "contract",
        symbol: "some_symbol",
        minter: senderAddress,
      },
      Math.random().toString(20).substring(2, 6),
      {
        gas: "200000",
        amount: [{ denom: "umlg", amount: "5000" }],
      }
    );

    return contractAddress;
  }

  async getOwnedTokens(ownerAddress: string): Promise<Array<Token>> {
    const ownedTokens = await this.signingClient.queryContractSmart(
      this.contractAddress,
      { tokens: { owner: ownerAddress } }
    );

    return ownedTokens;
  }

  async mintToken(
    senderAddress: string,
    owner: string,
    tokenId: string,
    tokenUri?: string
  ): Promise<string> {
    console.log(senderAddress, owner, this.contractAddress);
    const mintRes = await this.signingClient.execute(
      senderAddress,
      this.contractAddress,
      {
        mint: {
          token: {
            owner,
            token_id: tokenId,
            token_uri: tokenUri,
          },
        },
      },
      {
        gas: "200000",
        amount: [{ denom: "umlg", amount: "10000" }],
      }
    );

    return mintRes.transactionHash;
  }

  async transferToken(
    senderAddress: string,
    tokenId: string,
    receiver: string
  ): Promise<string> {
    const transferRes = await this.signingClient.execute(
      senderAddress,
      this.contractAddress,
      {
        transfer_nft: {
          token_id: tokenId,
          to: receiver,
        },
      },
      {
        gas: "200000",
        amount: [{ denom: "umlg", amount: "5000" }],
      }
    );

    return transferRes.transactionHash;
  }
}
