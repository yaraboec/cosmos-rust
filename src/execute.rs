use cosmwasm_std::{Binary, DepsMut, Env, MessageInfo, Response, StdResult};
use cw2::set_contract_version;
use cw721::Cw721ReceiveMsg;

use crate::{
    msg::{ExecuteMsg, InstantiateMsg, TokenMsg},
    response::ContractInfoResponse,
    state::{Contract, Token, CONTRACT_NAME, CONTRACT_VERSION},
    ContractError,
};

impl<'a> Contract<'a> {
    pub fn instantiate(
        &self,
        deps: DepsMut,
        _env: Env,
        info: MessageInfo,
        msg: InstantiateMsg,
    ) -> StdResult<Response> {
        set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;

        let contract_info = ContractInfoResponse {
            name: msg.name,
            symbol: msg.symbol,
        };
        let owner = info.sender;
        let minter = deps.api.addr_validate(&msg.minter)?;

        self.minter.save(deps.storage, &minter)?;
        self.owner.save(deps.storage, &owner)?;
        self.contract_info.save(deps.storage, &contract_info)?;

        Ok(Response::new().add_attribute("action", "init"))
    }

    pub fn execute(
        &self,
        deps: DepsMut,
        env: Env,
        info: MessageInfo,
        msg: ExecuteMsg,
    ) -> Result<Response, ContractError> {
        match msg {
            ExecuteMsg::Mint { token } => self.mint(deps, env, info, token.clone()),
            ExecuteMsg::SendNft {
                token_id,
                contract,
                msg,
            } => self.send_nft(deps, env, info, token_id, contract, msg),
            ExecuteMsg::TransferNft { token_id, to } => {
                self.transfer_nft(deps, env, info, token_id, to)
            }
        }
    }
}

impl<'a> Contract<'a> {
    pub fn mint(
        &self,
        deps: DepsMut,
        _env: Env,
        info: MessageInfo,
        msg: TokenMsg,
    ) -> Result<Response, ContractError> {
        if info.sender.clone() != self.minter.load(deps.storage)? {
            return Err(ContractError::Unauthorized {});
        }

        let token = Token {
            owner: deps.api.addr_validate(&msg.owner)?,
            token_id: msg.token_id.clone(),
            token_uri: msg.token_uri,
        };

        self.tokens
            .update(deps.storage, &msg.token_id, |old| match old {
                Some(_) => Err(ContractError::SomeError {}),
                None => Ok(token),
            })?;

        return Ok(Response::new().add_attribute("action", "mint"));
    }

    pub fn send_nft(
        &self,
        deps: DepsMut,
        env: Env,
        info: MessageInfo,
        token_id: String,
        contract: String,
        msg: Binary,
    ) -> Result<Response, ContractError> {
        self.internal_transfer(deps, env, info.clone(), &token_id, &contract)?;

        let send_msg = Cw721ReceiveMsg {
            sender: info.sender.into_string(),
            token_id,
            msg,
        };

        Ok(Response::new()
            .add_message(send_msg.into_cosmos_msg(contract.clone())?)
            .add_attribute("action", "transfer"))
    }

    pub fn transfer_nft(
        &self,
        deps: DepsMut,
        env: Env,
        info: MessageInfo,
        token_id: String,
        to: String,
    ) -> Result<Response, ContractError> {
        self.internal_transfer(deps, env, info, &token_id, &to)?;

        Ok(Response::new().add_attribute("action", "transfer"))
    }

    pub fn internal_transfer(
        &self,
        deps: DepsMut,
        _env: Env,
        info: MessageInfo,
        token_id: &str,
        to: &str,
    ) -> Result<Token, ContractError> {
        let mut token = self.tokens.load(deps.storage, token_id)?;

        if token.owner != info.sender {
            return Err(ContractError::Unauthorized {});
        }

        token.owner = deps.api.addr_validate(to)?;
        self.tokens.save(deps.storage, token_id, &token)?;

        return Ok(token);
    }
}
