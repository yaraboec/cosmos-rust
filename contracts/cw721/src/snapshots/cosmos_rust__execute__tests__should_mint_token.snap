---
source: src/execute.rs
expression: "contract.tokens.load(&deps.storage, TOKEN_ID).unwrap()"
---
{
  "owner": "owner",
  "token_id": "1",
  "token_uri": null
}
